src/System/TimedScopes/Correlation.cs

```diff
@@ -0,0 +1,302 @@
+					return existingContext.Data.TryGetValue(LoggingContextKey, out data) && data != null;
+				}
+
+				return false;
+			}
+			set
+			{
+				CorrelationData correlation = CurrentCorrelation;
+				if (correlation != null)
+				{
+					correlation.ShouldLogDirectly = value;
+				}
+				else
+				{
+					ICallContext existingContext = CallContext.ExistingCallContext();
+					if (existingContext != null)
+					{
+						if (value)
+						{
+							existingContext.Data[LoggingContextKey] = new object();
+						}
+						else
+						{
+							existingContext.Data.Remove(LoggingContextKey);
+						}
+					}
+				}
+			}
+		}
+
+
+		/// <summary>
+		/// The correlation handler used to store the correlation
+		/// </summary>
+		/// <remarks>
+		/// Calling methods on this interface bypasses the
+		/// event handlers for correlation and commits the correlation
+		/// directly to the correlation handler.
+		/// </remarks>
+		public ICorrelationStorage CorrelationHandler { get; private set; }
+
+
+		/// <summary>
+		/// The correlation handler used to store the correlation
+		/// </summary>
+		/// <remarks>
+		/// Calling methods on this interface bypasses the
+		/// event handlers for correlation and commits the correlation
+		/// directly to the correlation handler.
+		/// </remarks>
+		public static ICallContextManager CallContextManager { get; private set; }
+
+
+		/// <summary>
+		/// The correlation handler used to store the correlation
+		/// </summary>
+		/// <remarks>
+		/// Calling methods on this interface bypasses the
+		/// event handlers for correlation and commits the correlation
+		/// directly to the correlation handler.
+		/// </remarks>
+		public static IMachineInformation MachineInformation { get; private set; }
+
+
+		/// <summary>
+		/// Start a correlation
+		/// </summary>
+		/// <param name="data">correlation to set on the thread, null for default (new) correlation</param>
+		public void CorrelationStart(CorrelationData data)
+		{
+			data = CorrelationHandler.CorrelationStart(data);
+			if (data != null)
+			{
+				data.ParentCorrelation = CurrentCorrelation;
+				if (data.ParentCorrelation == null)
+				{
+					data.ShouldLogDirectly = ShouldLogDirectly;
+				}
+				CurrentCorrelation = data;
+
+				// Note: Creating a copy of the event handler to avoid multi-threaded race conditions
+				// Not creating extension methods to avoid unnecessary creating of arguments if not set
+				EventHandler<CorrelationEventArgs> correlationStarted = CorrelationStarted;
+				if (correlationStarted != null)
+				{
+					correlationStarted(this, new CorrelationEventArgs(data));
+				}
+			}
+		}
+
+
+		/// <summary>
+		/// Add data to the current correlation
+		/// </summary>
+		/// <remarks>If there is no current correlation, starts a new correlation</remarks>
+		/// <param name="key">key (name) of the correlation</param>
+		/// <param name="value">value of the added correlation data</param>
+		public void CorrelationAdd(string key, string value)
```

Comment:
>key [](start = 36, length = 3)

Validate?

Reply:
Both key and value are validated in CorrelationData.AddData() method. Should I do it also from higher level in CorrelationAdd method?